block-layer/content-addressable-archives.md

```diff
@@ -0,0 +1,150 @@
+ * A compressed integer prefixing each block, including the header block, indicating the total length of that block including CID
+
+This diagram shows how IPLD blocks, their root CID, and a header combine to form a CAR.
+
+![Content Addressable aRchive Diagram](content-addressable-archives.png)
+
+## Format Description
+
+The CAR format comprises a sequence of length-prefixed IPLD block data, where the first block in the CAR is the Header encoded as CBOR, and the remaining blocks form the Data component of the CAR and are each additionally prefixed with their CIDs. The length prefix of each block in a CAR is encoded as a "varint"&mdash;an unsigned [LEB128](https://en.wikipedia.org/wiki/LEB128) integer. This integer specifies the number of remaining bytes for that block entry&mdash;excluding the bytes used to encode the integer, but including the CID for non-header blocks.
+
+```
+|--------- Header --------| |---------------------------------- Data -----------------------------------|
+
+[ varint | dag-cbor block ] [ varint | CID | block ] [ varint | CID | block ] [ varint | CID | block ] …
+```
+
+### Header
+
+The first bytes of the CAR format hold a varint, this unsigned integer specifies the number of bytes beyond the varint itself that contain the _Header_ block. This Header block is a byte array DAG-CBOR (CBOR with tag 42 for CIDs) encoded object holding the version number and array of roots. As an [IPLD Schema](../schemas/):
+
+```ipldsch
+type CarHeader struct {
+  version Int
+  roots [&Any]
+}
+```
+
+#### Constraints
+
+* The `version` is always a value of `1`. Future iterations of this specification may make use of `version` to introduce variations of the format.
+* The `roots` array must contain **zero or more** CIDs, each of which must be present somewhere in the remainder of the CAR.
+
+An empty `roots` array is valid, but indicates that the CAR does not necessarily contain blocks forming a coherent DAG. Such a CAR may, for instance, form part of a collection of archives that, when combined, contain an entire DAG.
+
+### Data
+
+Immediately following the Header block, **zero or more** IPLD blocks are concatenated to form the _Data_ section of the CAR format. Each block is encoded into a _Section_ by the concatenation of the following values:
+
+1. Length in bytes of the combined CID and data in this Section, encoded as a varint
+2. CID of the block in this Section, encoded in the raw byte form of the CID
+3. Binary data of the block
+
+#### Length
+
+Each Section begins with a varint representation of an unsigned integer indicating the number of bytes containing the remainder of the section.
+
+#### CID
+
+Following the Length, the CID of the block is included in raw byte form. A decoder reading a Section must decode the CID according to CID byte encoding rules, which don't provide a stable length. See https://github.com/multiformats/cid for details on the encoding of a CID. CIDv0 and CIDv1 are currently supported.
```

Comment:
Do we need to support CIDv0 or do we support it because we can? I don't see a reason to use CIDv0 these days and I'd be happy to phase them out.

Reply:
We need to support CIDv0, we use it everywhere in IPFS. Long ago we could have probably introduced some kind of "use CIDv1 everywhere but always encode DagPB CIDs as CIDv0 when encoding a DagPB block" rule, but that ship has sailed.

Reply:
In order to create .car files you need some processing anyway. So doing a v0 to v1 conversion shouldn't be that hard.

Reply:
The issue is that forming a block full of CIDv1 CIDs produces a different CID than a block formed using CIDv0 CIDs. I'm concerned about erasing the CIDv1/v0 information.

But... this does bring up a good point. We might reference the same block as CIDv1 and CIDv0. Given that, I think you're right. It might make sense to specify that CID "headers" (i.e., not internal CIDs), must be normalized to CIDv1.

There's a second (unlikely) issue where multiple blocks might be valid under different codecs. This is one of the reasons we're changing our blockstores to refer to blocks by multihash, instead of by CID.

Reply:
CIDv0 is supported in current CAR code, but iirc the only public use of them is shipping FIL genesis blocks around so I guess it could be broken. Would need some consensus from folks responsible for go-car to make that happen.

Reply:
If we have a CIDv1 requirement on the headers we should make sure to document it well and require that CAR readers will accept a CIDv0 as the read input and convert to CIDv1 when looking in the .car file. This implementation detail is likely to be overlooked if not well documented.